modules/basegl/src/processors/instancerenderer.cpp

```diff
@@ -412,38 +428,150 @@ void InstanceRenderer::initializeResources() {
         prio += 100;
     }
 
+    std::array<const GLchar*, 1> feedbackVaryings = {"worldPosition"};
+    shader_.setTransformFeedbackVaryings(feedbackVaryings, GL_INTERLEAVED_ATTRIBS);
+
     shader_.build();
 }
 
-void InstanceRenderer::process() {
-    utilgl::activateTargetAndClearOrCopySource(outport_, background_);
+namespace {
 
-    if (vecPorts_.empty()) return;
+size_t numberOfVertices(DrawType dt, ConnectivityType ct, size_t nIndices) {
+    return util::numberOfVerticesForPrimitive(dt) * util::numberOfPrimitives(dt, ct, nIndices);
+}
+size_t numberOfVertices(Mesh::MeshInfo mi, size_t nIndices) {
+    return numberOfVertices(mi.dt, mi.ct, nIndices);
+}
+size_t numberOfVertices(const MeshGL& meshGL, size_t index) {
+    auto mi = meshGL.getMeshInfoForIndexBuffer(index);
+    auto nIndices = meshGL.getIndexBuffer(index)->getSize();
+    return numberOfVertices(mi, nIndices);
+}
+size_t numberOfVertices(const MeshGL& meshGL) {
+    const std::size_t numIndexBuffers = meshGL.getIndexBufferCount();
+    if (numIndexBuffers > 0) {
+        auto indices = std::ranges::iota_view{size_t{0}, numIndexBuffers};
+        return std::transform_reduce(indices.begin(), indices.end(), size_t{0}, std::plus<>{},
+                                     [&](size_t ib) { return numberOfVertices(meshGL, ib); });
+    } else {
+        auto nIndices = meshGL.getBufferGL(0)->getSize();
+        return numberOfVertices(meshGL.getDefaultMeshInfo(), nIndices);
+    }
+}
 
-    shader_.activate();
+auto primitiveMode(DrawType dt) {
+    switch (dt) {
+        case DrawType::Points:
+            return GL_POINTS;
+        case DrawType::Lines:
+            return GL_LINES;
+        case DrawType::Triangles:
+            return GL_TRIANGLES;
+        default:
+            return GL_TRIANGLES;
+    }
+}
 
-    utilgl::GlBoolState depthTest(GL_DEPTH_TEST, true);
-    utilgl::BlendModeState blendModeStateGL(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+size_t commonNumberOfElements(auto& ports) {
+    auto valid = ports | std::views::filter(
+                             [](const auto& port) -> bool { return port.size().has_value(); });
+    if (valid.empty()) return 0;
 
-    utilgl::setUniforms(shader_, camera_, lightingProperty_);
+    const auto minIt = std::ranges::min_element(valid, std::less<>{},
+                                                [](const auto& port) { return *port.size(); });
+    return *minIt->size();
+}
 
-    const auto& mesh = *inport_.getData();
-    MeshDrawerGL::DrawObject drawer{mesh.getRepresentation<MeshGL>(), mesh.getDefaultMeshInfo()};
-    utilgl::setShaderUniforms(shader_, mesh, "geometry");
+void forEach(const MeshGL& meshGL, Shader& shader, auto& ports, size_t nInstances,
+             auto&& callback) {
+    if (meshGL.empty()) return;
+
+    const std::size_t numIndexBuffers = meshGL.getIndexBufferCount();
+
+    for (size_t i = 0; i < nInstances; ++i) {
```

Comment:
Should the instance rendering be in the inner loop, just before the callback? This would reduce the number of buffer switches

Reply:
But it increases the number of uniform calls. effectively doing it that way results in about a 40% drop in fps